components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/OrientAsyncHelper.java

```diff
@@ -0,0 +1,132 @@
+import com.orientechnologies.orient.core.sql.query.OSQLAsynchQuery;
+import com.orientechnologies.orient.core.sql.query.OSQLNonBlockingQuery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Helper class to wrap Orient {@link OSQLAsynchQuery} into {@link Iterable}s.
+ *
+ * @since 3.0
+ */
+public class OrientAsyncHelper
+    extends ComponentSupport
+{
+  private static final int BUFFER_SIZE = 128;
+
+  private static final long TIMEOUT_SECONDS = 60L;
+
+  /**
+   * Executes with passed in {@link ODatabaseDocumentTx} connection the passed in {@link String} SELECT SQL command
+   * and returns a true asynchronous backed {@link Iterable} of {@link ODocument} instance, that can be later
+   * transformed into whatever caller needs. This means that this wrapper supports ANY SQL commands, even those
+   * performing aggregation on results, as long as iterator is not accumulated into a real array or list.
+   */
+  public static Iterable<ODocument> asyncIterable(final ODatabaseDocumentTx db,
+                                                  final String selectQuery,
+                                                  @Nullable final Map<String, Object> parameters)
+  {
+    checkNotNull(db);
+    checkNotNull(selectQuery);
+    final ArrayBlockingQueue<ODocument> queue = new ArrayBlockingQueue<>(BUFFER_SIZE);
+
+    db.command(new OSQLNonBlockingQuery<ODocument>(selectQuery, new OCommandResultListener()
+    {
+      @Override
+      public boolean result(final Object iRecord) {
+        final ODocument doc = (ODocument) iRecord;
+        try {
+          queue.offer(doc, TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          throw Throwables.propagate(e);
+        }
+        return true;
+      }
+
+      @Override
+      public void end() {
+        try {
+          queue.offer(SENTINEL, TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          throw Throwables.propagate(e);
+        }
+      }
+    })).execute(parameters);
+
+    return new OrientAsyncIterableWrapper(queue);
+  }
+
+  private static final ODocument SENTINEL = new ODocument();
+
+  private static final class OrientAsyncIterableWrapper
+      implements Iterable<ODocument>, Iterator<ODocument>
+  {
+    private final BlockingQueue<ODocument> queue;
+
+    private ODocument next;
+
+    public OrientAsyncIterableWrapper(final BlockingQueue<ODocument> queue)
+    {
+      this.queue = queue;
+    }
+
+    @Override
+    public Iterator<ODocument> iterator() {
+      return this;
+    }
+
+    @Override
+    public boolean hasNext() {
+      try {
+        next = queue.poll(TIMEOUT_SECONDS, TimeUnit.SECONDS);
+      }
+      catch (InterruptedException e) {
+        throw Throwables.propagate(e);
+      }
+      return next != SENTINEL;
+    }
+
+    @Override
+    public ODocument next() {
+      if (next == SENTINEL) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Use equals() to compare object references. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=pmd%3ACompareObjectsWithEquals)